File: demos/deadlock.js
/* Demo 5 — Deadlock.
 *
 * Two threads, two locks, opposite acquisition order.
 *  - Broken: A grabs lock1 then lock2; B grabs lock2 then lock1 → cycle.
 *  - Fixed:  Both grab lock1 first, then lock2.
 *
 * A 2.5-second watchdog stops the broken run so the page does not hang.
 */

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

class Mutex {
  constructor() { this.locked = false; this.waiters = []; }
  async acquire(signal) {
    while (this.locked) {
      if (signal?.aborted) throw new Error('abort');
      await new Promise((res) => {
        this.waiters.push(res);
        if (signal) signal.addEventListener('abort', res, { once: true });
      });
    }
    this.locked = true;
  }
  release() {
    this.locked = false;
    if (this.waiters.length) {
      const r = this.waiters.shift(); r();
    }
  }
}

async function attempt(send, fix, timeoutMs) {
  const lock1 = new Mutex();
  const lock2 = new Mutex();
  const state = { A: 'idle', B: 'idle' };

  function emit() { send({ type: 'tick', state: { ...state }, fix }); }
  function setState(who, v) { state[who] = v; emit(); }

  const ctrl = new AbortController();
  const start = performance.now();

  async function threadA() {
    setState('A', 'want lock1');
    await lock1.acquire(ctrl.signal);
    setState('A', 'holds lock1');
    await sleep(50);
    setState('A', 'want lock2');
    await lock2.acquire(ctrl.signal);
    setState('A', 'holds 1+2');
    await sleep(50);
    lock2.release();
    lock1.release();
    setState('A', 'done');
  }
  async function threadB_broken() {
    setState('B', 'want lock2');
    await lock2.acquire(ctrl.signal);
    setState('B', 'holds lock2');
    await sleep(50);
    setState('B', 'want lock1');
    await lock1.acquire(ctrl.signal);   // hangs
    setState('B', 'holds 1+2');
    await sleep(50);
    lock1.release();
    lock2.release();
    setState('B', 'done');
  }
  async function threadB_fixed() {
    setState('B', 'want lock1');
    await lock1.acquire(ctrl.signal);
    setState('B', 'holds lock1');
    await sleep(50);
    setState('B', 'want lock2');
    await lock2.acquire(ctrl.signal);
    setState('B', 'holds 1+2');
    await sleep(50);
    lock2.release();
    lock1.release();
    setState('B', 'done');
  }

  const watchdog = sleep(timeoutMs).then(() => {
    ctrl.abort();
    return 'timeout';
  });

  const a = threadA().catch(() => {});
  const b = (fix ? threadB_fixed() : threadB_broken()).catch(() => {});

  const winner = await Promise.race([
    Promise.all([a, b]).then(() => 'done'),
    watchdog,
  ]);

  return {
    deadlocked: winner === 'timeout',
    elapsed_ms: Math.round((performance.now() - start) * 10) / 10,
  };
}

export async function run(params, send) {
  const fix = !!params?.fix;
  const timeout = (params?.timeout_s ?? 2.5) * 1000;
  await send({ type: 'phase', label: fix ? 'Fixed order' : 'Broken order' });
  const r = await attempt(send, fix, timeout);
  await send({ type: 'result', ...r });
  await send({ type: 'done' });
}

File: legacy_python/app/demos/deadlock.py
"""
Demo 5 — Deadlock

What it shows:
  Two threads, two locks. Thread A grabs lock 1 then tries lock 2; Thread
  B grabs lock 2 then tries lock 1. They wait for each other forever.

Fix:
  Always acquire locks in the same global order. If both threads grab
  lock 1 first, no cycle can form.

Why it matters in RTOS:
  A real-time system that deadlocks is broken — the deadline is missed
  and there is no recovery. Order rules and lock hierarchies are how
  kernels avoid this.
"""

import asyncio
import threading
import time

from .base import Sender, make_bridge


def _attempt(bridge, fix: bool, timeout_s: float) -> dict:
    lock1 = threading.Lock()
    lock2 = threading.Lock()

    state = {"A": "idle", "B": "idle"}
    state_lock = threading.Lock()

    def set_state(who: str, value: str):
        with state_lock:
            state[who] = value
        bridge.emit({"type": "tick", "state": dict(state), "fix": fix})

    finished = threading.Event()

    def thread_a():
        set_state("A", "want lock1")
        lock1.acquire()
        set_state("A", "holds lock1")
        time.sleep(0.05)  # give B time to grab lock2
        set_state("A", "want lock2")
        # In FIX mode both threads also want lock1 first — but A already
        # has it, so this branch only happens on the broken version.
        lock2.acquire()
        set_state("A", "holds 1+2")
        time.sleep(0.05)
        lock2.release()
        lock1.release()
        set_state("A", "done")

    def thread_b_broken():
        set_state("B", "want lock2")
        lock2.acquire()
        set_state("B", "holds lock2")
        time.sleep(0.05)  # give A time to grab lock1
        set_state("B", "want lock1")
        lock1.acquire()           # ← will block forever vs A
        set_state("B", "holds 1+2")
        time.sleep(0.05)
        lock1.release()
        lock2.release()
        set_state("B", "done")

    def thread_b_fixed():
        # Same global order: lock1 first, lock2 second.
        set_state("B", "want lock1")
        lock1.acquire()
        set_state("B", "holds lock1")
        time.sleep(0.05)
        set_state("B", "want lock2")
        lock2.acquire()
        set_state("B", "holds 1+2")
        time.sleep(0.05)
        lock2.release()
        lock1.release()
        set_state("B", "done")

    target_b = thread_b_fixed if fix else thread_b_broken
    a = threading.Thread(target=thread_a, daemon=True)
    b = threading.Thread(target=target_b, daemon=True)
    a.start()
    b.start()

    start = time.monotonic()
    a.join(timeout=timeout_s)
    b.join(timeout=max(0.0, timeout_s - (time.monotonic() - start)))

    deadlocked = a.is_alive() or b.is_alive()
    return {
        "deadlocked": deadlocked,
        "elapsed_ms": round((time.monotonic() - start) * 1000.0, 1),
    }


async def run(send: Sender, params: dict) -> None:
    fix = bool(params.get("fix", False))
    timeout_s = float(params.get("timeout_s", 2.5))
    bridge = make_bridge(send)

    await send({"type": "phase", "label": "Fixed order" if fix else "Broken order"})
    result = await asyncio.to_thread(_attempt, bridge, fix, timeout_s)
    await send({"type": "result", **result})
    await send({"type": "done"})
